File: MultiLanguageSandbox/src/test/deps/r/15.R



# This function 'calculate_game_points' calculates the total points scored in a series of game rounds. In each round, a player is given n integers. The player scores a point in a round if they can select two numbers from these integers whose sum equals k. The function takes the total number of rounds n, the target sum k, and a list of vectors, where each vector represents the integers given in each round.
# Examples:
# calculate_game_points(3, 20, list(c(1, 3, 5), c(9, 15, 5), c(1, 10, 10))): This returns 2, as in two of the rounds, the player can find pairs that sum up to 20.


calculate_game_points <- function(n, k, game_rounds) {
  total_points <- 0
  
  for (round in game_rounds) {
    found <- FALSE
    for (i in 1:(length(round) - 1)) {
      for (j in (i + 1):length(round)) {
        if (round[i] + round[j] == k) {
          found <- TRUE
          break
        }
      }
      if (found) break
    }
    if (found) {
      total_points <- total_points + 1
    }
  }
  
  return(total_points)
}
# Test cases
main <- function() {
    stopifnot(calculate_game_points(3, 20, list(c(1, 3, 5), c(9, 15, 5), c(1, 10, 10))) == 2)
    stopifnot(calculate_game_points(2, 10, list(c(2, 5, 3), c(1, 9, 4))) == 1)
    stopifnot(calculate_game_points(4, 15, list(c(4, 11, 7), c(2, 6, 9), c(8, 5, 2), c(3, 12, 6))) == 3)
    stopifnot(calculate_game_points(3, 8, list(c(1, 2, 5), c(3, 4, 1), c(6, 2, 0))) == 1)
    stopifnot(calculate_game_points(2, 5, list(c(1, 1, 1), c(2, 2, 1))) == 0)
}

# Run tests
main()

File: MultiLanguageSandbox/src/test/deps/dart/15.dart

import 'dart:math';

int getMaxValue(int totalMoney, int numItems, List<int> prices, List<int> importance)
/*Calculates the maximum value one can obtain with a given amount of money and a list of items with associated prices and importance.

This function uses a dynamic programming approach to determine the maximum value that can be achieved
within the constraints of the total money available to spend. Each item's value is determined by its
price multiplied by its importance. The function iterates over the list of items, updating the maximum
value obtainable at each price point up to the total money available.

Arguments:
  totalMoney (int): The total amount of money available to spend on items.
  numItems (int): The number of different items available for purchase.
  prices (List<int>): A list containing the price of each item.
  importance (List<int>): A list containing the importance of each item, corresponding to the prices list.

Returns:
  int: The maximum value obtainable with the given total money and items.

Examples:
    getMaxValue(50, 2, [20, 30], [5, 10])
    returns 400
*/
{
  // Create a DP array where dp[i] represents the maximum value achievable with i money
  List<int> dp = List.filled(totalMoney + 1, 0);

  for (int i = 0; i < numItems; i++) {
    int price = prices[i];
    int value = price * importance[i];
    
    // Iterate from the end to avoid reusing the same item multiple times
    for (int j = totalMoney; j >= price; j--) {
      if (dp[j - price] + value > dp[j]) {
        dp[j] = dp[j - price] + value;
      }
    }
  }

  return dp[totalMoney];
}
void testGetMaxValue() {
  // Test case 1: Basic scenario with two items
  assert(
    getMaxValue(50, 2, [20, 30], [5, 10]) == 400,
    'Test Case 1 Failed: Expected maximum value to be 60, got ${getMaxValue(50, 2, [20, 30], [5, 10])}'
  );

  // Test case 2: Scenario where the total money can't buy the most valuable item
  assert(
    getMaxValue(10, 3, [5, 10, 15], [3, 6, 9]) == 60,
    'Test Case 2 Failed: Expected maximum value to be 6, got ${getMaxValue(10, 3, [5, 10, 15], [3, 6, 9])}'
  );

  // Test case 3: Multiple items can be bought within the total money
  assert(
    getMaxValue(100, 4, [50, 25, 10, 20], [1, 2, 3, 2]) == 140,
    'Test Case 3 Failed: Expected maximum value to be 100, got ${getMaxValue(100, 4, [50, 25, 10, 20], [1, 2, 3, 2])}'
  );

  // Additional test cases can be added here as needed.

  print('All test cases passed!');
}

void main() {
  testGetMaxValue();
}